File: framework/go/dispatch.go
package dispatch

import (
	"context"
	"fmt"
)

// RuneFunc is the signature for any registered callable.
// Input and output are JSON-encoded bytes for isolation.
type RuneFunc func(ctx context.Context, input []byte) ([]byte, error)

// Middleware wraps a callable invocation.
type Middleware func(name string, next RuneFunc) RuneFunc

// Dispatcher is an immutable callable registry and call router.
type Dispatcher struct {
	runes      map[string]RuneFunc
	middleware []Middleware
}

// New creates an immutable Dispatcher with the given runes and middleware.
func New(runes map[string]RuneFunc, middleware []Middleware) *Dispatcher {
	return &Dispatcher{runes: runes, middleware: middleware}
}

// Call invokes a callable by name through the middleware chain.
func (d *Dispatcher) Call(ctx context.Context, name string, input []byte) ([]byte, error) {
	fn, ok := d.runes[name]
	if !ok {
		return nil, fmt.Errorf("callable %q not registered", name)
	}

	for i := len(d.middleware) - 1; i >= 0; i-- {
		fn = d.middleware[i](name, fn)
	}

	return fn(ctx, input)
}

File: framework/ts/dispatch.ts
export type RuneFunc = (input: string) => Promise<string>;
export type Middleware = (name: string, next: RuneFunc) => RuneFunc;

export class Dispatcher {
  private readonly runes: Map<string, RuneFunc>;
  private readonly middleware: Middleware[];

  constructor(runes: Map<string, RuneFunc>, middleware: Middleware[]) {
    this.runes = new Map(runes);
    this.middleware = [...middleware];
  }

  async call(name: string, input: string): Promise<string> {
    let fn = this.runes.get(name);
    if (!fn) {
      throw new Error(`callable "${name}" not registered`);
    }
    for (let i = this.middleware.length - 1; i >= 0; i--) {
      fn = this.middleware[i](name, fn);
    }
    return fn(input);
  }
}
